feat: unflag responsive images (#13917)
withastro/astro: .changeset/mean-gifts-know.md

---
'astro': minor
'@astrojs/vercel': patch
---

> Responsive images are now stable.

Reviewer: Sorry I'm not used to unflagging feature so I hadn't noticed that before, but shouldn't this changeset include the version were the feature was introduced? I was looking at [Astro Docs Docs](https://contribute.docs.astro.build/docs-for-code-changes/experimental-feature-docs/#unflagging-an-experimental-feature), so maybe something like:
```suggestion
The responsive images feature introduced behind a flag in [v5.0.0](https://github.com/withastro/astro/blob/main/packages/astro/CHANGELOG.md#500) is no longer experimental and is available for general use.
```

The `experimental.responsiveImages` flag has been removed, and all experimental image configuration options have been renamed to their final names.

> ## Migration Guide

Reviewer 2: Before this, don't forget this is now hype time for this "new feature"! Feel free to pull in stuff from the original experimental PR, describing what responsive images are, why they're cool etc.!

```diff
export default defineConfig({
  image: {
+    responsiveStyles: true,
  },
});
```

The experimental image configuration options have been renamed:

**Before:**
```js
export default defineConfig({
  image: {
    experimentalLayout: 'constrained',
    experimentalObjectFit: 'cover', 
    experimentalObjectPosition: 'center',
    experimentalBreakpoints: [640, 750, 828, 1080, 1280],
    experimentalDefaultStyles: true, 
  },
  experimental: {
    responsiveImages: true,
  },
});
```

**After:**
```js
export default defineConfig({
  image: {
    layout: 'constrained',
    objectFit: 'cover',
    objectPosition: 'center', 
    breakpoints: [640, 750, 828, 1080, 1280],
    responsiveStyles: true, // This is now *false* by default
  },
});
```

### Component usage remains the same

The `layout`, `fit`, and `position` props on `<Image>` and `<Picture>` components work exactly the same as before:

```astro
<Image 
  src={myImage} 
  alt="A responsive image"
  layout="constrained"
  fit="cover"
  position="center"
/>
```

> If you weren't using the experimental responsive images feature, no changes are required.

Reviewer 2: ```suggestion
If you weren't using the experimental responsive images feature, no changes are required.

Please see the [Image guide](https://docs.astro.build/en/guides/images/#responsive-image-behavior) for more information on using responsive images in Astro.
```
Tentatively, that's the link for responsive image content!

The responsive images feature introduced behind a flag in [v5.0.0](https://github.com/withastro/astro/blob/main/packages/astro/CHANGELOG.md#500) is no longer experimental and is available for general use.

> Displaying images correctly on the web can be challenging, and is one of the most common performance issues seen in sites. While the venerable `<img>` tag makes it easy to display a basic image, for the best performance you need to understand a panoply of attributes and tags. While the Astro image components have helped to simplify this process, generating optimized, lazy-loaded images, you have still been left with the most challenging part: responsive images.
> 
> The new responsive images feature in Astro provides a simple way to handle this, automatically generating optimized images for different screen sizes and resolutions, and applying the correct attributes to ensure that images are displayed correctly on all devices. 
> 
> ## Usage
> 
> To enable responsive images, set the `image.layout` option in the Astro config, or the `layout` attribute on the `<Image>` or `<Picture>` components. You should also enable the `image.responsiveStyles` option in your Astro config. 
> 
> For full details, see the new [Image guide](https://docs.astro.build/en/guides/images/#responsive-image-behavior).

Reviewer 2: ```suggestion
The new responsive images feature in Astro automatically generates optimized images for different screen sizes and resolutions, and applies the correct attributes to ensure that images are displayed correctly on all devices. 

Enable the `image.responsiveStyles` option in your Astro config. Then, set a `layout` attribute on any <Image /> or <Picture /> component, or configure a default `image.layout`, for instantly responsive images with automatically generated `srcset` and `sizes` attributes based on the image's dimensions and the layout type.

Displaying images correctly on the web can be challenging, and is one of the most common performance issues seen in sites. This new feature simplifies the most challenging part of the process: serving your site visitor an image optimized for their viewing experience, and for your website's performance.

For full details, see the updated [Image guide](https://docs.astro.build/en/guides/images/#responsive-image-behavior).
```
Like in the other changeset, I think maybe frontloading the feature *is* the hype? I snagged something from the experimental release changeset to see how we initially presented the feature. What about a structure like this?

> ## Migration Guide

Reviewer 2: ```suggestion
## Migration from Experimental Responsive Images
```
I think it's worth being clear that this is about the steps to take if you previously were using experimental, and not e.g. "How to migrate your existing images to responsive images"
